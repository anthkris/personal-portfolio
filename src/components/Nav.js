import React, { useState, useCallback } from 'react';
import { Link } from 'gatsby';
import styled from 'styled-components';
import { useTrail, animated } from 'react-spring';
import { Turn as Hamburger } from './Turn';
import useBoop from '../utils/useBoop';
import logo from '../assets/images/knanthony_logo.svg';

const NavStyles = styled.nav`
  margin-bottom: 3rem;
  font-family: 'Fredoka One';
  font-size: 1rem;
  display: flex;
  justify-content: space-between;
  align-items: center;
  flex-wrap: wrap;
  button {
    background: transparent;
    border: none;
  }
  .mobileNavHidden {
    display: none;
  }
  .mobileNav {
    display: block;
    list-style: none;
    margin: 0;
    padding: 0;
    text-align: right;
  }
  .desktopNav {
    display: none;
  }
  li {
    order: 1;
    padding: 0.5rem 0px;
  }
  a {
    text-decoration: none;
    color: var(--black);
    padding: 0.3rem 0;
    &:hover,
    &:focus,
    &.active {
      background: var(--link-pink);
      color: #fff;
    }
  }
  .logoLink {
    &:hover,
    &:focus {
      background: transparent;
    }
  }
  .logo {
    max-width: 50px;
    width: 100%;
    min-width: 40px;
  }
  .trails-main {
    position: relative;
    overflow: hidden;
    cursor: pointer;
    flex-basis: 100%;
    font-size: 1.4rem;
  }

  .trails-text {
    position: relative;
    width: 100%;
    will-change: transform, opacity;
    overflow: hidden;
  }

  .trails-text > div {
    overflow: hidden;
  }
  /* Media screen size */
  @media screen and (min-width: 768px) {
    button {
      display: none;
    }
    .mobileNav,
    .trails-main {
      display: none;
    }
    .desktopNav {
      display: flex;
      text-align: center;
      list-style: none;
      align-items: center;
      margin-top: --6rem;
      flex-basis: initial;
      font-size: initial;
    }
    li {
      padding: 0px 5px;
    }
  }
`;

const Trail = ({ open, children, ...props }) => {
  const items = React.Children.toArray(children);
  const trail = useTrail(items.length, {
    config: { mass: 5, tension: 2000, friction: 200 },
    opacity: open ? 1 : 0,
    x: open ? 0 : 20,
    height: open ? 30 : 0,
    from: { opacity: 0, x: 20, height: 0 },
  });
  return (
    <div className="trails-main" {...props}>
      <ul className={open ? 'mobileNav' : 'mobileNavHidden'}>
        {trail.map(({ x, height, ...rest }, index) => (
          <animated.li
            key={`trail-li-${index}`}
            className='trails-text'
            style={{
              ...rest,
              transform: x.interpolate((x) => `translate3d(0,${x}px,0)`),
            }}
          >
            <animated.span style={{ height }}>{items[index]}</animated.span>
          </animated.li>
        ))}
      </ul>
    </div>
  );
};

const Nav = () => {
  const [menuIsOpen, setMenuIsOpen] = useState(false);
  const [style, trigger] = useBoop({ x: 2, y: 0, rotation: 0, scale: 1.2 });

  const toggleNav = useCallback(() => {
    setMenuIsOpen((menuIsOpen) => !menuIsOpen);
  }, []);
  return (
    <NavStyles>
      <Link className="logoLink" to="/" onMouseEnter={trigger}>
        <animated.img
          style={style}
          className="logo"
          src={logo}
          alt="Home page"
        />
      </Link>
      <button
        onClick={toggleNav}
        aria-haspopup='true'
        aria-label='Open or Close the menu'
        aria-expanded={menuIsOpen ? 'true' : 'false'}
      >
        <Hamburger toggled={menuIsOpen} />
      </button>
      <Trail open={menuIsOpen}>
        <Link activeClassName='active' to='/writing'>
          Writing
        </Link>
        <Link activeClassName="active" to="/play">
          Play
        </Link>
        <Link activeClassName="active" to="/explorables">
          Explorables
        </Link>
        <Link activeClassName="active" to="/contact">
          Contact
        </Link>
      </Trail>
      <ul className='desktopNav'>
        <li>
          <Link activeClassName='active' to='/writing'>
            Writing
          </Link>
        </li>
        <li>
          <Link activeClassName="active" to="/play">
            Play
          </Link>
        </li>
        <li>
          <Link activeClassName="active" to="/explorables">
            Explorables
          </Link>
        </li>
        <li>
          <Link activeClassName="active" to="/contact">
            Contact
          </Link>
        </li>
      </ul>
    </NavStyles>
  );
};

export default Nav;
