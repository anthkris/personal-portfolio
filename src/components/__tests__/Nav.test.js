import * as React from 'react';
import { act, render, screen } from '@testing-library/react';
import matchMediaPolyfill from 'mq-polyfill';
import '../../../__mocks__/matchMedia-mock';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import Nav from '../Nav';

// Test that desktopNav shows on desktop and mobileNav on mobile
// Snapshot info: https://medium.com/simply/snapshots-painless-testing-of-react-components-6bce3c4d51fc

describe('Nav', () => {
  it('renders the desktopNav top nav on desktop', async () => {
    const tree = renderer.create(<Nav />).toJSON();
    const { container } = render(<Nav />);
    // console.log(container.firstChild);
  });
  it('renders the mobileNav top nav on mobile', async () => {});
});
