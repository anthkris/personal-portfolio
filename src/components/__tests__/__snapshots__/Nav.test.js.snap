// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav renders the desktopNav top nav on desktop 1`] = `
.c0 {
  margin-bottom: 3rem;
  font-family: 'Fredoka One';
  font-size: 1rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c0 button {
  background: transparent;
  border: none;
}

.c0 .mobileNavHidden {
  display: none;
}

.c0 .mobileNav {
  display: block;
  list-style: none;
  margin: 0;
  padding: 0;
  text-align: right;
}

.c0 .desktopNav {
  display: none;
}

.c0 li {
  -webkit-order: 1;
  -ms-flex-order: 1;
  order: 1;
  padding: 0.5rem 0px;
}

.c0 a {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: var(--black);
  padding: 0.3rem 0;
}

.c0 a:hover,
.c0 a:focus,
.c0 a.active {
  background: var(--link-pink);
  color: #fff;
}

.c0 .logoLink:hover,
.c0 .logoLink:focus {
  background: transparent;
}

.c0 .logo {
  max-width: 50px;
  width: 100%;
  min-width: 40px;
}

.c0 .trails-main {
  position: relative;
  overflow: hidden;
  cursor: pointer;
  -webkit-flex-basis: 100%;
  -ms-flex-preferred-size: 100%;
  flex-basis: 100%;
  font-size: 1.4rem;
}

.c0 .trails-text {
  position: relative;
  width: 100%;
  will-change: transform,opacity;
  overflow: hidden;
}

.c0 .trails-text > div {
  overflow: hidden;
}

@media screen and (min-width:768px) {
  .c0 button {
    display: none;
  }

  .c0 .mobileNav,
  .c0 .trails-main {
    display: none;
  }

  .c0 .desktopNav {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    text-align: center;
    list-style: none;
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    margin-top: --6rem;
    -webkit-flex-basis: initial;
    -ms-flex-preferred-size: initial;
    flex-basis: initial;
    font-size: initial;
  }

  .c0 li {
    padding: 0px 5px;
  }
}

<nav
  className="c0"
>
  <a
    className="logoLink"
    href="/"
    onMouseEnter={[Function]}
  >
    <img
      alt="Home page"
      className="logo"
      src="test-file-stub"
      style={Object {}}
    />
  </a>
  <button
    aria-expanded="false"
    aria-haspopup="true"
    aria-label="Open or Close the menu"
    onClick={[Function]}
    type="button"
  >
    <div
      aria-hidden="true"
      className="hamburger-react"
      data-testid="turn"
      style={
        Object {
          "cursor": "pointer",
          "height": "48px",
          "outline": "none",
          "position": "relative",
          "transition": "0.4s cubic-bezier(0, 0, 0, 1)",
          "userSelect": "none",
          "width": "48px",
        }
      }
    >
      <div
        data-testid="burger-bars"
        style={
          Object {
            "background": "currentColor",
            "height": "3px",
            "left": "8px",
            "position": "absolute",
            "top": "13px",
            "transform": "none",
            "transition": "0.4s cubic-bezier(0, 0, 0, 1)",
            "width": "32px",
          }
        }
      />
      <div
        style={
          Object {
            "background": "currentColor",
            "height": "3px",
            "left": "8px",
            "position": "absolute",
            "top": "23px",
            "transform": "none",
            "transition": "0.2s cubic-bezier(0, 0, 0, 1)",
            "width": "32px",
          }
        }
      />
      <div
        style={
          Object {
            "background": "currentColor",
            "height": "3px",
            "left": "8px",
            "position": "absolute",
            "top": "33px",
            "transform": "none",
            "transition": "0.4s cubic-bezier(0, 0, 0, 1)",
            "width": "32px",
          }
        }
      />
    </div>
  </button>
  <div
    className="trails-main"
    data-testid="mobileNav"
  >
    <ul
      className="mobileNavHidden"
    >
      <li
        className="trails-text"
        style={
          Object {
            "opacity": 0,
            "transform": "translate3d(0,20px,0)",
          }
        }
      >
        <span
          style={
            Object {
              "height": 0,
            }
          }
        >
          <a
            href="/"
          >
            Home
          </a>
        </span>
      </li>
      <li
        className="trails-text"
        style={
          Object {
            "opacity": 0,
            "transform": "translate3d(0,20px,0)",
          }
        }
      >
        <span
          style={
            Object {
              "height": 0,
            }
          }
        >
          <a
            href="/writing"
          >
            Writing
          </a>
        </span>
      </li>
      <li
        className="trails-text"
        style={
          Object {
            "opacity": 0,
            "transform": "translate3d(0,20px,0)",
          }
        }
      >
        <span
          style={
            Object {
              "height": 0,
            }
          }
        >
          <a
            href="/play"
          >
            Play
          </a>
        </span>
      </li>
      <li
        className="trails-text"
        style={
          Object {
            "opacity": 0,
            "transform": "translate3d(0,20px,0)",
          }
        }
      >
        <span
          style={
            Object {
              "height": 0,
            }
          }
        >
          <a
            href="/about"
          >
            About
          </a>
        </span>
      </li>
    </ul>
  </div>
  <ul
    className="desktopNav"
    data-testid="desktopNav"
  >
    <li>
      <a
        href="/"
      >
        Home
      </a>
    </li>
    <li>
      <a
        href="/writing"
      >
        Writing
      </a>
    </li>
    <li>
      <a
        href="/play"
      >
        Play
      </a>
    </li>
    <li>
      <a
        href="/about"
      >
        About
      </a>
    </li>
  </ul>
</nav>
`;
